components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java

```diff
@@ -256,64 +301,171 @@ private void preSetUserClaimOnMobileNumberUpdate(Map<String, String> claims, Use
-            temporary claim exists as 'true' in the claim list. If 'UseVerifyClaim' is disabled, no need to
-            check for 'verifyMobile' claim.
-             */
-            if (Utils.isUseVerifyClaimEnabled() && !isVerifyMobileClaimAvailable(claims)) {
+            Mobile numbers in verifiedMobileNumbers claim are already verified. No need to verify again.
+            */
+            if (exisitingVerifiedNumbersList != null && exisitingVerifiedNumbersList.contains(mobileNumber)) {
                 Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
-                        .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_INAPPLICABLE_CLAIMS.toString());
+                        .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_ALREADY_VERIFIED_MOBILE_NUMBERS.toString());
                 invalidatePendingMobileVerification(user, userStoreManager, claims);
                 return;
+            } else {
+                String existingMobileNumber;
+                String username = user.getUserName();
+                try {
+                    existingMobileNumber = userStoreManager.getUserClaimValue(username, IdentityRecoveryConstants.
+                            MOBILE_NUMBER_CLAIM, null);
+                } catch (UserStoreException e) {
+                    String error = String.format("Error occurred while retrieving existing mobile number for user: %s in " +
+                            "domain: %s and user store: %s", username, user.getTenantDomain(), user.getUserStoreDomain());
+                    throw new IdentityEventException(error, e);
+                }
+
+                if (StringUtils.equals(mobileNumber, existingMobileNumber)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("The mobile number to be updated: %s is same as the existing mobile " +
+                                        "number for user: %s in domain: %s and user store: %s. Hence an SMS OTP " +
+                                        "verification will not be triggered.", mobileNumber, username,
+                                user.getTenantDomain(), user.getUserStoreDomain()));
+                    }
+                    Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
+                            .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_EXISTING_MOBILE_NUM.toString());
+                    invalidatePendingMobileVerification(user, userStoreManager, claims);
+
+                    if (supportMultipleMobileNumbers) {
+                        if (exisitingVerifiedNumbersList != null && !exisitingVerifiedNumbersList.contains(
+                                mobileNumber)) {
+                            exisitingVerifiedNumbersList.add(mobileNumber);
+                            claims.put(IdentityRecoveryConstants.VERIFIED_MOBILE_NUMBERS_CLAIM, String.join(",",
```

Comment:
Use the multi attribute separator instead of the hard coded ,

Reply:
We can use `FrameworkUtils.getMultiAttributeSeparator()`.

https://github.com/wso2/carbon-identity-framework/blob/4b4d6bf2b10c9a3fe913c73911d185718b625bdd/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java#L2254